File: ADSUS_Mobile/lib/features/ai_chatbot/data/dtos/chat_stream_event.dart
import 'dart:async';
import 'dart:convert';

/// Kiểu sự kiện Server-Sent Events (SSE) từ pipeline Chatbot AI.
sealed class ChatStreamEvent {
  const ChatStreamEvent();

  /// Parse sự kiện từ tên event và chuỗi data.
  factory ChatStreamEvent.parse(String eventType, String rawData) {
    final trimmedData = rawData.trim();
    Map<String, dynamic>? jsonMap;
    if (trimmedData.startsWith('{') && trimmedData.endsWith('}')) {
      try {
        final decoded = jsonDecode(trimmedData);
        if (decoded is Map<String, dynamic>) {
          jsonMap = decoded;
        }
      } catch (_) {
        // Fallback sang plain text nếu JSON malformed
      }
    }

    final normalizedEvent = eventType.trim().toLowerCase();

    // 1. Thinking event
    if (normalizedEvent == 'thinking' ||
        (normalizedEvent.isEmpty && jsonMap != null && jsonMap['status'] == 'thinking')) {
      final status = jsonMap != null ? (jsonMap['status'] as String? ?? 'thinking') : 'thinking';
      return ChatStreamThinkingEvent(status: status);
    }

    // 2. Delta event
    if (normalizedEvent == 'delta' ||
        (normalizedEvent.isEmpty && jsonMap != null && jsonMap.containsKey('chunk'))) {
      final chunk = jsonMap != null ? (jsonMap['chunk'] as String? ?? '') : rawData;
      return ChatStreamDeltaEvent(chunk);
    }

    // 3. Done event
    if (normalizedEvent == 'done' ||
        (normalizedEvent.isEmpty &&
            jsonMap != null &&
            (jsonMap.containsKey('messageId') || jsonMap.containsKey('content')))) {
      if (jsonMap != null) {
        return ChatStreamDoneEvent.fromJson(jsonMap);
      }
      return ChatStreamDoneEvent(
        messageId: '',
        role: 'ASSISTANT',
        content: rawData,
        createdAt: DateTime.now(),
      );
    }

    // 4. Error event
    if (normalizedEvent == 'error' ||
        (normalizedEvent.isEmpty &&
            jsonMap != null &&
            (jsonMap.containsKey('code') || jsonMap.containsKey('error')))) {
      final message = jsonMap != null
          ? (jsonMap['message'] as String? ?? jsonMap['error'] as String? ?? rawData)
          : rawData;
      final code = jsonMap != null ? jsonMap['code'] as String? : null;
      return ChatStreamErrorEvent(message, code: code);
    }

    // Fallback: nếu event không rõ nhưng có data, coi là delta text
    if (rawData.isNotEmpty) {
      if (jsonMap != null && jsonMap.containsKey('chunk')) {
        return ChatStreamDeltaEvent(jsonMap['chunk'] as String? ?? '');
      }
      return ChatStreamDeltaEvent(rawData);
    }

    return const ChatStreamThinkingEvent();
  }
}

/// Sự kiện model đang suy nghĩ / xử lý pre-checks.
class ChatStreamThinkingEvent extends ChatStreamEvent {
  const ChatStreamThinkingEvent({this.status = 'thinking'});

  final String status;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ChatStreamThinkingEvent && runtimeType == other.runtimeType && status == other.status;

  @override
  int get hashCode => status.hashCode;

  @override
  String toString() => 'ChatStreamThinkingEvent(status: $status)';
}

/// Sự kiện từng chunk text được LLM sinh ra.
class ChatStreamDeltaEvent extends ChatStreamEvent {
  const ChatStreamDeltaEvent(this.chunk);

  final String chunk;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ChatStreamDeltaEvent && runtimeType == other.runtimeType && chunk == other.chunk;

  @override
  int get hashCode => chunk.hashCode;

  @override
  String toString() => 'ChatStreamDeltaEvent(chunk: $chunk)';
}

/// Sự kiện kết thúc luồng streaming, chứa toàn bộ metadata và message đã lưu DB.
class ChatStreamDoneEvent extends ChatStreamEvent {
  const ChatStreamDoneEvent({
    required this.messageId,
    required this.role,
    required this.content,
    required this.createdAt,
    this.isSafetyResponse = false,
    this.detectedIntent,
    this.isRateLimitExceeded = false,
  });

  final String messageId;
  final String role;
  final String content;
  final DateTime createdAt;
  final bool isSafetyResponse;
  final String? detectedIntent;
  final bool isRateLimitExceeded;

  factory ChatStreamDoneEvent.fromJson(Map<String, dynamic> json) {
    DateTime parsedDate;
    if (json['createdAt'] != null) {
      parsedDate = DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now();
    } else {
      parsedDate = DateTime.now();
    }

    return ChatStreamDoneEvent(
      messageId: json['messageId'] as String? ?? '',
      role: json['role'] as String? ?? 'ASSISTANT',
      content: json['content'] as String? ?? '',
      createdAt: parsedDate,
      isSafetyResponse: json['isSafetyResponse'] as bool? ?? false,
      detectedIntent: json['detectedIntent'] as String?,
      isRateLimitExceeded: json['isRateLimitExceeded'] as bool? ?? false,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ChatStreamDoneEvent &&
          runtimeType == other.runtimeType &&
          messageId == other.messageId &&
          role == other.role &&
          content == other.content &&
          isSafetyResponse == other.isSafetyResponse &&
          detectedIntent == other.detectedIntent &&
          isRateLimitExceeded == other.isRateLimitExceeded;

  @override
  int get hashCode => Object.hash(
        messageId,
        role,
        content,
        isSafetyResponse,
        detectedIntent,
        isRateLimitExceeded,
      );

  @override
  String toString() =>
      'ChatStreamDoneEvent(id: $messageId, role: $role, intent: $detectedIntent, safety: $isSafetyResponse, limit: $isRateLimitExceeded)';
}

/// Sự kiện lỗi trong quá trình streaming.
class ChatStreamErrorEvent extends ChatStreamEvent {
  const ChatStreamErrorEvent(this.message, {this.code});

  final String message;
  final String? code;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ChatStreamErrorEvent &&
          runtimeType == other.runtimeType &&
          message == other.message &&
          code == other.code;

  @override
  int get hashCode => Object.hash(message, code);

  @override
  String toString() => 'ChatStreamErrorEvent(message: $message, code: $code)';
}

/// Helper phân giải byte stream thành `Stream<ChatStreamEvent>`.
///
/// Xử lý phân tách chunk UTF-8 (multibyte tiếng Việt) bằng `utf8.decoder`
/// và ghép line bị ngắt packet bằng `LineSplitter`.
Stream<ChatStreamEvent> parseSseStream(Stream<List<int>> byteStream) async* {
  String currentEvent = '';
  final dataBuffer = StringBuffer();

  final lineStream = byteStream
      .cast<List<int>>()
      .transform(utf8.decoder)
      .transform(const LineSplitter());

  await for (final line in lineStream) {
    if (line.isEmpty) {
      // Dòng trống kết thúc 1 frame SSE
      final dataStr = dataBuffer.toString();
      if (currentEvent.isNotEmpty || dataStr.isNotEmpty) {
        yield ChatStreamEvent.parse(currentEvent, dataStr);
      }
      currentEvent = '';
      dataBuffer.clear();
      continue;
    }

    if (line.startsWith(':')) {
      // Comment hoặc SSE ping/keep-alive
      continue;
    }

    if (line.startsWith('event:')) {
      currentEvent = line.substring(6).trim();
    } else if (line.startsWith('data:')) {
      final value = line.substring(5).trimLeft();
      if (dataBuffer.isNotEmpty) {
        dataBuffer.write('\n');
      }
      dataBuffer.write(value);
    }
  }

  // Flush nếu stream đóng mà không có dòng trống cuối
  final remainingData = dataBuffer.toString();
  if (currentEvent.isNotEmpty || remainingData.isNotEmpty) {
    yield ChatStreamEvent.parse(currentEvent, remainingData);
  }
}

File: ADSUS_BE.BLL/Engagement/Models/ChatStreamEvent.cs
using System.Text.Json.Serialization;
using ADSUS_BE.BLL.Engagement.DTOs;
using ADSUS_BE.BLL.Engagement.Services;
using ADSUS_BE.DAL.Entities;

namespace ADSUS_BE.BLL.Engagement.Models;

/// <summary>
/// Base class cho các Server-Sent Event (SSE) trong AI Chatbot pipeline.
/// </summary>
public abstract record ChatStreamEvent
{
    [JsonIgnore]
    public abstract string EventType { get; }
}

/// <summary>
/// Event "thinking" phát ra ngay khi các pre-checks (validation, rate limit, safety filter) thành công,
/// trước khi gọi LLM client.
/// Wire format: event: thinking\ndata: {"status":"thinking"}\n\n
/// </summary>
public record ChatThinkingEvent(string Status = "thinking") : ChatStreamEvent
{
    [JsonIgnore]
    public override string EventType => "thinking";

    [JsonPropertyName("status")]
    public string Status { get; init; } = Status;
}

/// <summary>
/// Event "delta" phát ra mỗi khi nhận được một chunk văn bản từ LLM client.
/// Wire format: event: delta\ndata: {"chunk":"..."}\n\n
/// </summary>
public record ChatDeltaEvent(string Chunk) : ChatStreamEvent
{
    [JsonIgnore]
    public override string EventType => "delta";

    [JsonPropertyName("chunk")]
    public string Chunk { get; init; } = Chunk;

    [JsonIgnore]
    public string Text => Chunk;
}

/// <summary>
/// Event "done" phát ra khi kết thúc stream (hoặc phát ngay lập tức khi safety/rate limit chặn).
/// Chứa thông tin hoàn chỉnh của assistant message đã được lưu trong DB.
/// Wire format: event: done\ndata: {"messageId":"...","content":"...","detectedIntent":"...","isSafetyResponse":...}\n\n
/// </summary>
public record ChatDoneEvent : ChatStreamEvent
{
    [JsonIgnore]
    public override string EventType => "done";

    [JsonPropertyName("messageId")]
    public Guid MessageId { get; init; }

    [JsonPropertyName("content")]
    public string Content { get; init; } = string.Empty;

    [JsonPropertyName("detectedIntent")]
    public string? DetectedIntent { get; init; }

    [JsonPropertyName("isSafetyResponse")]
    public bool IsSafetyResponse { get; init; }

    [JsonPropertyName("isRateLimitExceeded")]
    public bool IsRateLimitExceeded { get; init; }

    [JsonPropertyName("createdAt")]
    public DateTime CreatedAt { get; init; }

    [JsonPropertyName("role")]
    public string Role { get; init; } = "Assistant";

    [JsonIgnore]
    public ChatMessageResponse Response => new()
    {
        MessageId = MessageId,
        Content = Content,
        Role = Enum.TryParse<ChatRole>(Role, true, out var role) ? role : ChatRole.Assistant,
        CreatedAt = CreatedAt,
        IsSafetyResponse = IsSafetyResponse,
        IsRateLimitExceeded = IsRateLimitExceeded,
        DetectedIntent = Enum.TryParse<ChatIntent>(DetectedIntent, true, out var parsedIntent) ? parsedIntent : null
    };

    public ChatDoneEvent() { }

    public ChatDoneEvent(
        Guid messageId,
        string content,
        string? detectedIntent,
        bool isSafetyResponse,
        bool isRateLimitExceeded = false,
        DateTime? createdAt = null,
        string role = "Assistant")
    {
        MessageId = messageId;
        Content = content;
        DetectedIntent = detectedIntent;
        IsSafetyResponse = isSafetyResponse;
        IsRateLimitExceeded = isRateLimitExceeded;
        CreatedAt = createdAt ?? DateTime.UtcNow;
        Role = role;
    }

    public ChatDoneEvent(ChatMessageResponse response)
    {
        MessageId = response.MessageId;
        Content = response.Content;
        Role = response.Role.ToString();
        CreatedAt = response.CreatedAt;
        IsSafetyResponse = response.IsSafetyResponse;
        DetectedIntent = response.DetectedIntent?.ToString();
        IsRateLimitExceeded = response.IsRateLimitExceeded;
    }
}

/// <summary>
/// Event "error" phát ra khi xảy ra lỗi trong quá trình streaming.
/// Wire format: event: error\ndata: {"message":"...","code":"STREAM_ERROR"}\n\n
/// </summary>
public record ChatErrorEvent(string Message, string? Code = "STREAM_ERROR") : ChatStreamEvent
{
    [JsonIgnore]
    public override string EventType => "error";

    [JsonPropertyName("message")]
    public string Message { get; init; } = Message;

    [JsonPropertyName("code")]
    public string? Code { get; init; } = Code;
}

// Aliases for alternate naming conventions
public sealed record ThinkingStreamEvent : ChatThinkingEvent
{
    public ThinkingStreamEvent(string status = "thinking") : base(status) { }
}

public sealed record DeltaStreamEvent : ChatDeltaEvent
{
    public DeltaStreamEvent(string chunk) : base(chunk) { }
}

public sealed record DoneStreamEvent : ChatDoneEvent
{
    public DoneStreamEvent(ChatMessageResponse response) : base(response) { }
    public DoneStreamEvent(
        Guid messageId,
        string content,
        string? detectedIntent,
        bool isSafetyResponse,
        bool isRateLimitExceeded = false,
        DateTime? createdAt = null,
        string role = "Assistant")
        : base(messageId, content, detectedIntent, isSafetyResponse, isRateLimitExceeded, createdAt, role) { }
}

public sealed record ErrorStreamEvent : ChatErrorEvent
{
    public ErrorStreamEvent(string message, string? code = "STREAM_ERROR") : base(message, code) { }
}
